vmware/ngc/NGC-AdapterManager/src/main/java/org/opensds/vmware/ngc/adapters/oceanstor/OceanStor.java

```diff
@@ -0,0 +1,506 @@
+        String name = jsonObject.getString("NAME");
+        String id = jsonObject.getString("ID");
+        String wwn = jsonObject.getString("WWN");
+        ALLOC_TYPE allocType = (jsonObject.getInt("ALLOCTYPE") == 1) ? ALLOC_TYPE.THIN : ALLOC_TYPE.THICK;
+        long capacity = jsonObject.getLong("CAPACITY") * 512;
+
+        return new VolumeMO(name, id, wwn, allocType, capacity);
+    }
+}
+
+class StoragePoolMOBuilder {
+    static public StoragePoolMO build(JSONObject jsonObject) {
+        String name = jsonObject.getString("NAME");
+        String id = jsonObject.getString("ID");
+        POOL_TYPE type = (jsonObject.getInt("USAGETYPE") == 1) ? POOL_TYPE.BLOCK : POOL_TYPE.FILE;
+        long totalCapacity = jsonObject.getLong("USERTOTALCAPACITY") * 512;
+        long freeCapacity = jsonObject.getLong("USERFREECAPACITY") * 512;
+
+        return new StoragePoolMO(name, id, type, totalCapacity, freeCapacity);
+    }
+}
+
+public class OceanStor extends Storage {
+    class Handler implements Request.RequestHandler {
+        @Override
+        public void setRequestBody(HttpEntityEnclosingRequestBase req, Object body) {
+            StringEntity entity = new StringEntity(body.toString(), "utf-8");
+            req.setEntity(entity);
+        }
+
+        @Override
+        public Object parseResponseBody(String body) throws JSONException {
+            return new JSONObject(body);
+        }
+    }
+
+    RestClient client;
+
+    public OceanStor(String name) {
+        super(name);
+        this.client = new RestClient();
+    }
+
+    public void login(String ip, int port, String user, String password) throws Exception {
+        client.login(ip, port, user, password);
+    }
+
+    public void logout() {
+        client.logout();
+    }
+
+    public StorageMO getDeviceInfo() throws Exception {
+        JSONObject system = client.getSystem();
+
+        String name = system.getString("NAME");
+        String model = "unknown";
+        if (system.has("productModeString")) {
+            model = system.getString("productModeString");
+        } else if (system.has("PRODUCTMODE")) {
+            model = ProductModel.getModel(system.getString("PRODUCTMODE"));
+        }
+
+        String sn = system.getString("wwn");
+        String status = RunningStatus.getStatus(system.getString("RUNNINGSTATUS"));
+
+        return new StorageMO(name, model, sn, status,"Huawei");
+    }
+
+    public VolumeMO createVolume(String name, String description, ALLOC_TYPE allocType, long capacity, String poolId) throws Exception {
+        JSONObject volume = client.createVolume(name, allocType, capacity, poolId);
+        return VolumeMOBuilder.build(volume);
+    }
+
+    public void deleteVolume(String volumeId) throws Exception {
+        client.deleteVolume(volumeId);
+    }
+
+    public List<VolumeMO> listVolumes() throws Exception {
+        List<VolumeMO> volumes = new ArrayList<>();
+
+        JSONArray jsonArray = client.listVolumes("");
```

Comment:
Hi，we do not have logs. Because we hope to make a thin adapter layer. Error message will be collected by plugins.  And if  jsonArray == NULL , it will return a list of volumes which the length is 0.

Reply:
Ok